### PR TITLE
Feature Minio S3 Enhancement

### LINKED ISSUE #34: Data transfer: extend samples to utilize S3 (minio)
## WHAT

The MXD walkthrough should be extended so that one provider connector, e.g. "Alice" hosts an asset, that is actually stored in S3 (using the MinIO container image). 
Bob then transfers the data into some arbitrary location, e.g. his own S3 store.

In the course of this issue, it should also be tested, whether a transfer HTTP -> S3 can be achieved.

## WHY
Highlighting standard Tx-EDC use cases and making them accessible for users of MXD.

## HOW
- [ ] add terraform resources for MinIO (_suggestion_: use a `kubernetes_deployment`)
- [ ] create a bucket on provider side and put some test data into that bucket. This needs to be done automatically, without manual commands. (_suggestion_: could use a `job`)
- [ ] add relevant documentation to `mxd/README.md` outlining the process, including the correct `curl` command and transfer request JSON
- [ ] repeat for transfers `HTTP -> S3`
- [ ] Suggestion: after each transfer, show the contents of the relevant database tables on provider and consumer side, to further illustrate what's going on.

### PR DESCRIPTION
**Pull request by SantoshDeepak1**
Feature Minio S3 committed

## Description
This pull request (PR) introduces the integration of Minio's S3 capabilities with Eclipse EDC, enabling the transfer of assets from a source S3 provider to a target S3 consumer.

Closes #34

### CODE REVIEW

**Comment by SantoshDeepak1** [comment]
## Current Issue

Currently, after a successful transfer of assets from the source to the target, an exception is being logged in the source dataplane.
```
Failed to send callback request
org.eclipse.edc.spi.http.EdcHttpClientException: Server response to Request{method=POST, url=http://localhost:8083/control/transferprocess/18d14e52-7796-4e01-add3-7f25802352d5/complete} was not one of [200, 204] but was 405: 
```
I attempted to override callback address to the transfer payload as described ("MXD Management API_S3_MINIO.json"), but it seems to have no effect, as the system is still using a default callback address.

File: mxd/README.md [hunk 1]
```diff
@@ -974,7 +974,216 @@ That's because, Trudy had created the [access policy](#762-create-access-policy)
...
+        "type": "AmazonS3",
+        "name": "document_alice",
+        "id": "document_alice",
+        "contenttype": "text/plain"
+    },
+    "dataAddress": {
+        "@type": "DataAddress",
+        "secretAccessKey": "qwerty123",
+        "accessKeyId": "qwerty123",
+        "endpointOverride": "http://10.96.103.86:9000",
```
Comment (hemantxpatel) [inline mxd/README.md:1013]: Is `endpointOverride` needed. It is already defined as environment variable.

File: mxd/README.md [hunk 2]
```diff
@@ -974,7 +974,216 @@ That's because, Trudy had created the [access policy](#762-create-access-policy)
...
+
+#### 8.3 Create Contract Definition
+Create a contract definition using the following `curl` command:
+```shell
+curl --location 'http://localhost/alice/management/v2/contractdefinitions' \
+--header 'Content-Type: application/json' \
+--header 'X-Api-Key: password' \
+--data-raw '{
+    "@context": {},
+    "@id": "2",
```
Comment (hemantxpatel) [inline mxd/README.md:1067]: St this point contract definition with `id: 2` already exist. Use another one.

File: mxd/README.md [hunk 3]
```diff
@@ -974,7 +974,216 @@ That's because, Trudy had created the [access policy](#762-create-access-policy)
...
+--header 'X-Api-Key: password' \
+--data-raw '{
+	"@context": {
+		"odrl": "http://www.w3.org/ns/odrl/2/"
+	},
+	"@type": "NegotiationInitiateRequestDto",
+	"connectorAddress": "http://alice-controlplane:8084/api/v1/dsp",
+	"protocol": "dataspace-protocol-http",
+	"connectorId": "BPNL000000000001",
+    "callbackAddresses": null,
```
Comment (hemantxpatel) [inline mxd/README.md:1095]: Fix indentation.

File: mxd/minio.tf [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    template {
+      metadata {
+        name = "create-minio-bucket"
+      }
+      spec {
+        container {
+          name    = "minio-client"
+          image   = "minio/mc:latest" # Use the appropriate MinIO client image
+          command = ["/bin/sh", "-c"]
+          args    = ["mc config host add minio http://${local.minio-url} ${local.minio-username} ${local.minio-password} && mc mb minio/alice-bucket"]
```
Comment (hemantxpatel) [inline mxd/minio.tf:157]: Can we combine both bucket creation job in a single job. `mc mb minio/alice-bucket && mc mb minio/bob-bucket`

File: mxd/minio.tf [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+#  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#  License for the specific language governing permissions and limitations
+#  under the License.
+#
+#  SPDX-License-Identifier: Apache-2.0
+#
+
+
+resource "kubernetes_namespace" "minio" {
```
Comment (paullatzelsperger) [inline mxd/minio.tf:21]: don't create a separate namespace for this

File: mxd/minio.tf [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+  data = {
+    (local.minio-seed_collection_name) = file("./assets/testdocument.txt")
+  }
+}
+
+locals {
+  minio-seed_collection_name = "testdocument.txt"
+  minio-ip                   = "10.96.103.86"
+  minio-port                 = "9000"
+  minio-username             = "qwerty123"
```
Comment (paullatzelsperger) [inline mxd/minio.tf:257]: username and pwd should be `var`s, so they can be configured by the user.

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
the minio deployment should be a terraform `module` rather than a static config that contains hardocded `bob` and `alice`. Much like the `connector` module is done.

It could even be a module, that is used in turn by the `connector` module

File: mxd/README.md (hunk 3 above)
Comment (SantoshDeepak1) [reply mxd/README.md:1095]: Fixed.

File: mxd/README.md (hunk 2 above)
Comment (SantoshDeepak1) [reply mxd/README.md:1067]: Id 3 Used

File: mxd/README.md (hunk 1 above)
Comment (SantoshDeepak1) [reply mxd/README.md:1013]: Removed endpointOverride.

**Comment by SantoshDeepak1** [comment]
Hello @paullatzelsperger , could you provide more details regarding your comment? 
**Should the deployment of Minio be structured as a Terraform module, similar to how the connector module has been implemented?** 
In the current setup, there is a single Minio server where two distinct buckets are created, one for Alice and one for Bob. Are you suggesting that we should have separate Minio servers for each connector, with each server hosting assets for different participants, such as Alice and Bob?

**Comment by paullatzelsperger** [comment]
@SantoshDeepak1 Like I said: for every connector there should be a separate MinIO deployment (~server) that hosts the bucket(s). This should be implemented as terraform module, so that the same module declaration can be used for bob and alice. please use appropriate variables to distinguish them.

> Are you suggesting that we should have separate Minio servers for each connector, with each server hosting assets for different participants, such as Alice and Bob?

this is what i'm suggesting.

**Comment by SantoshDeepak1** [comment]
Yes, @paullatzelsperger, opting for MinIO providers makes the process more straightforward. However, in issue #34, you suggested utilizing kubernetes_deployment. Currently, we are shifting our strategy to incorporate terraform providers.

**Comment by paullatzelsperger** [comment]
> Yes, @paullatzelsperger, opting for MinIO providers makes the process more straightforward. However, in issue #34, you suggested utilizing kubernetes_deployment. Currently, we are shifting our strategy to incorporate terraform providers.

either one is fine i think, so long as it is re-usable

**Comment by SantoshDeepak1** [comment]
Hello @paullatzelsperger, I've implemented the code based on your recommendations. Could you please review it and let me know if I'm heading in the correct direction?

File: mxd/minio.tf (hunk 4 above)
Comment (SantoshDeepak1) [reply mxd/minio.tf]: As the Minio servers are now separated into their own modules, the task of creating buckets will occur within their respective modules.

**Comment by SantoshDeepak1** [comment]
Hello @paullatzelsperger ,

I have incorporated all your suggestions; please review the changes. According to @hemantxpatel , there is no requirement to include accessKeyId, secretAccessKey, and endpointOverride in the payload . Based on my observations, these parameters are not necessary when creating assets on the provider side. However, these parameters become necessary when a consumer requests to transfer assets from the provider; otherwise, it leads to a NoSuchBucketException. Kindly examine the code and advise if this behavior is intentional or if adjustments are needed in the payload.

While creating asset from the provider the payload is 
```
{
    "@context": {},
    "asset": {
        "@type": "Asset",
        "@id": "3",
        "properties": {
            "description": "Product EDC Demo Asset 3- S3"
        }
    },
    "properties": {
        "version": "1.0",
        "type": "AmazonS3",
        "name": "document_alice",
        "id": "document_alice",
        "contenttype": "text/plain"
    },
    "dataAddress": {
        "@type": "DataAddress",
        "type": "AmazonS3",
        "blobname": "document.txt",
        "region":"us-east-1",
        "bucketName": "alice-bucket",
        "keyName":"document.txt"
    }
}
```
While transfer request from consumer the payload is 

```
{
    "@context": {
       "@vocab": "https://w3id.org/edc/v0.0.1/ns/"
    },
    "assetId": "3",
    "connectorAddress": "{{ALICE_PROVIDER_PROTOCOL_DSP}}",
    "connectorId": "BPNL000000000002",
    "contractId": "37e5556e-c43c-4d7a-941e-21d40284af0f",
    "dataDestination": {
        "type": "AmazonS3",
        "keyName":"document.txt",
        "bucketName": "bob-bucket",
        "region":"us-east-1",
        "endpointOverride":"http://10.244.0.38:9000",
        "accessKeyId":"qwerty123",
        "secretAccessKey":"qwerty123"
    },
    "protocol": "dataspace-protocol-http"
}
```

**Comment by paullatzelsperger** [comment]
@SantoshDeepak1 Please take a look at the [EDC Samples](https://github.com/eclipse-edc/Samples/tree/main/transfer/transfer-05-file-transfer-cloud) to see what data must be included when creating assets, and when requesting a transfer. The sample does create assets on Azure Blob, but AWS S3 works analogously

**Comment by SantoshDeepak1** [comment]
Hi @paullatzelsperger,

Thanks for sharing the resource. While examining the transfer payload in the sample, I noticed that it doesn't explicitly mention the accessKeyId and secretAccessKey. During the creation of the connector using the helm chart, I override the following section, and the provider utilizes these values when pulling from the S3 store. However, the transfer consumer encounters difficulty storing these values in their own S3 store without explicitly specifying accessKeyId and secretAccessKey. The example you provided uses the AWS CLI and stores them in a vault. Therefore, I'm seeking a method to transmit these secrets to the connector via a vault.

In the connector module main.tf file line no 65

```
dataplane: {
  aws: {
    endpointOverride: "http://${var.minio-config.minio-url}",
    accessKeyId: var.minio-config.minio-username,
    secretAccessKey: var.minio-config.minio-password
   }
}
```

**Comment by SantoshDeepak1** [comment]
Hello @paullatzelsperger ,

I've made an attempt using the examples provided in the resources you shared. However, our situation involves Minio being recreated after each Terraform destroy and apply cycle, posing challenges in setting a temporary session token or access key in the vault. This issue results in the following exception:

```software.amazon.awssdk.services.s3.model.S3Exception: The Access Key Id you provided does not exist in our records```

 Minio doesn't support AWS temporary session tokens. There are discussions and issues raised regarding this limitation:

[GitHub Issue 2444](https://github.com/minio/mc/issues/2444)

The workaround mentioned in the ticket involves editing the mc configuration, but attempts to do so resulted in a "permission denied" error due to it being a read-only file.

Can you please suggest.

**Comment by hemantxpatel** [comment]
I have refactored the code. Some of the major changes includes:
- Instead of creating `minIO` separately for `Alice` and `Bob`, connector module uses minIO module to deploy a MinIO instance for itself.
- Moved tutorial from main `README.md` to a file under `docs` folder.
- Current tutorial requires actual AWS access key / secret during initiate transfer due to MinIO limitations. I have raised an issue #169 to fix it in case we have some workaround or MinIO adds support.

**Comment by hemantxpatel** [comment]
I messed up this branch as I was trying to create a branch with both azurite and S3 changes. Will open a new PR.

**Pull request closed without merging**